operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDiscoveryService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.operator.v2alpha1;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.IntOrString;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.ServiceBuilder;
+import io.fabric8.kubernetes.api.model.ServiceSpec;
+import io.fabric8.kubernetes.api.model.ServiceSpecBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.OperatorManagedResource;
+import org.keycloak.operator.StatusUpdater;
+import org.keycloak.operator.v2alpha1.crds.Keycloak;
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusBuilder;
+
+import java.util.Optional;
+
+public class KeycloakDiscoveryService extends OperatorManagedResource implements StatusUpdater<KeycloakStatusBuilder> {
+
+    private Service existingService;
+
+    public KeycloakDiscoveryService(KubernetesClient client, Keycloak keycloakCR) {
+        super(client, keycloakCR);
+        fetchExistingService().ifPresent(s -> this.existingService = s);
+    }
+
+    private ServiceSpec getServiceSpec() {
+      return new ServiceSpecBuilder()
+              .addNewPort()
+              .withPort(Constants.KEYCLOAK_DISCOVERY_SERVICE_PORT)
+              .endPort()
+              .withSelector(Constants.DEFAULT_LABELS)
+              .withClusterIP("None")
+              .build();
+    }
+
+    @Override
+    protected Optional<HasMetadata> getReconciledResource() {
+        return Optional.of(
+                Optional.ofNullable(existingService)
+                    .map(a -> {
+                            a.setSpec(getServiceSpec());
+                            return a;
+                            })
+                .orElse(newService()));
```

Comment:
Again, feels a bit overengineered and hard to read. I know we had something similar in the [original PR]() but this is even more complicated.

But just a nit pick, not a blocker.

Reply:
moved to plain `null` checks